### PR TITLE
Synchronize parallel calls to Conn.Close and Conn.handshake

### PR DESCRIPTION
**Pull request by Danielius1922**
#### Reference issue
Fixes #...

### CODE REVIEW

**Comment by Sean-Der** [comment]
Overall LGTM @Danielius1922!

Looks like a different race was found in CI. Once you can get it all to go green happy to test and can fix up any nits myself :)

**Comment by Danielius1922** [comment]
> Overall LGTM @Danielius1922!
> 
> Looks like a different race was found in CI. Once you can get it all to go green happy to test and can fix up any nits myself :)

Yeah the issue is that `c.handshakeLoopsFinished.Wait()` (from Conn.Close) and `c.handshakeLoopsFinished.Add(2)` (from Conn.handshake) are called from different goroutines. Since `Close` can be called from a different goroutine, the calls on `c.handshakeLoopsFinished` need to be synchronized. 
I think the idea in `Conn.Close` is that it calls the internal `Conn.close` method, which might interrupt an ongoing handshake and then we wait using `c.handshakeLoopsFinished.Wait()`, so there is no ongoing handshake after the call to `Conn.Close`.

Handshakes themselves are serialized using `Conn.handshakeMutex`. I could try locking it in `Conn.Close`, I'm afraid of causing a deadlock if `Conn.handshake` on some code path invokes `Conn.Close`. 
Better safe than sorry here I think, so I'm gonna introduce a channel, assign/get it under a lock and wait on it outside the lock.

File: conn.go [hunk 1]
```diff
@@ -405,7 +411,12 @@ func (c *Conn) Write(p []byte) (int, error) {
 // Close closes the connection.
 func (c *Conn) Close() error {
 	err := c.close(true) //nolint:contextcheck
-	c.handshakeLoopsFinished.Wait()
+	c.closeLock.Lock()
```
Comment (Danielius1922) [inline conn.go:414]: poor `c.closeLock` gets locked and unlocked thrice in a call to `c.Close` (twice in `close` + the call here), I guess I could get the channel in `c.close` and return it? Doesn't fit semantically for `close` to return the channel, but it is an internal function, so it doesn't matter much.

File: conn.go [hunk 2]
```diff
@@ -1146,8 +1161,13 @@ func (c *Conn) translateHandshakeCtxError(err error) error {
...
 
 func (c *Conn) close(byUser bool) error {
-	c.cancelHandshaker()
-	c.cancelHandshakeReader()
+	c.closeLock.Lock()
+	cancelHandshaker := c.cancelHandshaker
+	cancelHandshakeReader := c.cancelHandshakeReader
+	c.closeLock.Unlock()
+
+	cancelHandshaker()
```
Comment (Danielius1922) [inline conn.go:1169]: @Sean-Der is it important to invoke these cancels at the beginning? if not I could move thes copying to line 1179 (after closeLock gets locked) and the invocations to 1187 (after it gets unlocked)

Comment (Danielius1922) [reply conn.go:1169]: Tests succeed locally when I do it like that, so maybe the order of calls here isn't imporant, in which case I can do it like that and avoid multiple locking/unlocking of the lock.

**Review by Sean-Der — APPROVED** [review]

**Comment by Sean-Der** [comment]
Tagged `v3.0.2`

@Danielius1922 I added you to the org! Feel free to make branches and tags.

I think the concurrency of `Conn` can be cleaned up a lot. If you want to take a hack at it I am fully in support :)

**Pull request merged**